File: erpnext/student_services/report/student_feedback_analysis/student_feedback_analysis.js
// Copyright (c) 2026, Frappe Technologies Pvt. Ltd. and contributors
// For license information, please see license.txt

frappe.query_reports["Student Feedback Analysis"] = {
	"filters": [
		{
			"fieldname": "feedback_type",
			"label": __("Feedback Type"),
			"fieldtype": "Link",
			"options": "Feedback Type",
			"width": 150
		},
		{
			"fieldname": "college",
			"label": __("College"),
			"fieldtype": "Link",
			"options": "Company",
			"width": 150
		}
	],
	
	"formatter": function(value, row, column, data, default_formatter) {
		// Color code for average ratings
		if (column.fieldname === "average_rating") {
			if (value >= 4.0) {
				return `<span style="color: green; font-weight: bold;">${parseFloat(value).toFixed(2)}</span>`;
			} else if (value >= 3.0) {
				return `<span style="color: orange;">${parseFloat(value).toFixed(2)}</span>`;
			} else if (value < 2.0) {
				return `<span style="color: red; font-weight: bold;">${parseFloat(value).toFixed(2)}</span>`;
			}
			return `<span>${parseFloat(value).toFixed(2)}</span>`;
		}
		
		// Highlight high student participation
		if (column.fieldname === "student_count" && value > 50) {
			return `<span style="color: blue; font-weight: bold;">${value}</span>`;
		}
		
		return default_formatter(value, row, column, data);
	},
	
	// Add chart configuration
	"get_chart_data": function(columns, data) {
		if (!data || data.length === 0) {
			return null;
		}
		
		// Prepare data for chart - Top 10 questions by student count
		let sorted_data = [...data].sort((a, b) => b.student_count - a.student_count);
		let top_10 = sorted_data.slice(0, 10);
		
		let labels = top_10.map(d => {
			let question = d.question;
			return question.length > 40 ? question.substring(0, 37) + "..." : question;
		});
		
		let values = top_10.map(d => d.student_count);
		
		return {
			data: {
				labels: labels,
				datasets: [
					{
						name: __("Number of Students"),
						values: values,
						chartType: "bar"
					}
				]
			},
			type: "bar",
			title: __("Top 10 Questions by Student Participation"),
			height: 300,
			colors: ["#4CAF50"]
		};
	},
	
	"onload": function(report) {
		// Add export buttons
		report.page.add_inner_button(__("Export to Excel"), function() {
			report.export_report("xlsx");
		});
		
		report.page.add_inner_button(__("Export to CSV"), function() {
			report.export_report("csv");
		});
	}
};

File: erpnext/student_services/report/student_feedback_analysis/student_feedback_analysis.py
# Copyright (c) 2024, Your Company and contributors
# For license information, please see license.txt

import frappe
from frappe import _

def execute(filters=None):
    """Main method to execute the report"""
    
    # Define columns
    columns = [
        {
            "fieldname": "feedback_type",
            "label": _("Feedback Type"),
            "fieldtype": "Link",
            "options": "Feedback Type",
            "width": 150
        },
        {
            "fieldname": "college",
            "label": _("College"),
            "fieldtype": "Link",
            "options": "Company",
            "width": 200
        },
        {
            "fieldname": "question",
            "label": _("Question"),
            "fieldtype": "Data",
            "width": 400
        },
        {
            "fieldname": "student_count",
            "label": _("Number of Students"),
            "fieldtype": "Int",
            "width": 150
        },
        {
            "fieldname": "average_rating",
            "label": _("Average Rating"),
            "fieldtype": "Float",
            "precision": 2,
            "width": 150
        },
        {
            "fieldname": "total_responses",
            "label": _("Total Responses"),
            "fieldtype": "Int",
            "width": 150
        }
    ]
    
    # Build conditions based on filters
    conditions = "1=1"
    
    if filters:
        if filters.get("feedback_type"):
            conditions += f" AND sf.feedback_type = '{filters.get('feedback_type')}'"
        
        if filters.get("college"):
            conditions += f" AND sf.college = '{filters.get('college')}'"
    
    # SQL query - Count unique students per question
    query = f"""
        SELECT 
            sf.feedback_type,
            sf.college,
            r.question,
            COUNT(DISTINCT sf.student) AS student_count,
            AVG(r.rating) AS average_rating,
            COUNT(r.rating) AS total_responses
        FROM 
            `tabStudent Feedback` sf
        INNER JOIN 
            `tabRating Responses` r ON sf.name = r.parent
        WHERE 
            {conditions}
        GROUP BY 
            sf.feedback_type,
            sf.college, 
            r.question
        ORDER BY 
            sf.feedback_type ASC,
            student_count DESC
    """
    
    data = frappe.db.sql(query, as_dict=1)
    
    # Prepare chart data for single chart
    chart = get_chart_data(data)
    
    # Return with chart
    return columns, data, None, chart

def get_chart_data(data):
    """Prepare chart data for visualization"""
    
    if not data:
        return None
    
    # Chart 1: Student Count by Question (Top 10)
    sorted_by_students = sorted(data, key=lambda x: x.get('student_count', 0), reverse=True)
    top_questions = sorted_by_students[:10]
    
    # Prepare labels and values for chart
    labels = []
    values = []
    
    for d in top_questions:
        # Truncate long question text
        question_text = d.get('question', '')
        if len(question_text) > 40:
            question_text = question_text[:37] + '...'
        labels.append(question_text)
        values.append(d.get('student_count', 0))
    
    chart = {
        "data": {
            "labels": labels,
            "datasets": [
                {
                    "name": _("Number of Students"),
                    "values": values
                }
            ]
        },
        "type": "bar",
        "title": _("Top 10 Questions by Student Participation"),
        "height": 300,
        "colors": ["#4CAF50"]
    }
    
    return chart